### DISCUSSION TITLE
Compiler can't determine best common type for `switch` expression

### DISCUSSION THREAD

**Discussion by gurustron** [dotnet/csharplang#8841] [Q&A]
Probably my understanding of the spec in this part is incorrect, but according to [`switch` expression feature spec](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/patterns#switch-expression):

> The type of the switch_expression **is the best common type** ([§11.6.3.15](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/expressions.md#116315-finding-the-best-common-type-of-a-set-of-expressions)) of the expressions appearing to the right of the => tokens of the switch_expression_arms if such a type exists and the expression in every arm of the switch expression can be implicitly converted to that type.

And the [natural function type](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/lambda-improvements#natural-function-type) has the following:

> Function_types are used in a few specific contexts only:
> - implicit and explicit conversions
> - method type inference ([§11.6.3](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/expressions.md#1163-type-inference)) and **best common type** ([§11.6.3.15](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/expressions.md#116315-finding-the-best-common-type-of-a-set-of-expressions))
> - var initializers

And:

> Best common type ([§11.6.3.15](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/expressions.md#116315-finding-the-best-common-type-of-a-set-of-expressions)) is defined in terms of type inference so the type inference changes above apply to best common type as well.

With following sample which compiles:

```
var fs = new[] { (string s) => s.Length, (string s) => int.Parse(s) };
```

So for array construction compiler is able to determine the best common type.

**Version Used**: 
.NET 8.0.101, C# 12

**Steps to Reproduce**:

Create a `swtich` expression with anonymous lambdas:
```
var func = (args.Length == 1) switch
{
    false => () => "Hello World",
    true => () => "Foo Bar"
};
```

A minimal repro, with source-code provided, is ideal.  [Repro @sharplab](https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgBuBDATgAgGYDOOAvDgHYCmA7gNoC6OA3jgBQCUJAfDgEQASFCBAD2PADSsOxbjwDqwrBAAmPHAF8A3DgD02nDACsAHhgIADBLBlgneqlSZceAK5kAxiVbYA5gQB0ADIUZN7AABYkpAgcBFRgwG5hqIyoOGn4GBAEFFySufyCIjjyiipiqenAWM450nl1PABiwsI4AELYPKiaQA==)

**Expected Behavior**:

Compiles

**Actual Behavior**:

> error CS8506: No best type was found for the switch expression.

**Comment by PoseidonEnergy** [comment]
I would also like to highlight another discrepancy related to functions.

This does not compile:

    var str = (() => "Hello World")(); // does not compile

However, this compiles:

    var str = ((Func<string>)(() => "Hello World"))(); // compiles

**Comment by RikkiGibson** [comment]
@cston for visibility. I think this is a subtle issue relating to lambda natural type.

**Comment by jaredpar** [comment]
@cston ping for comment on whether or not this is about lambda natural type

**Comment by cston** [comment]
It looks like the *best common type* calculation for switch expressions currently ignores *function types*. 

There is a similar issue with conditional expressions:
```csharp
// error CS0173: Type of conditional expression cannot be determined
var func = (args.Length == 1) ? () => "Hello" : () => "World";
```

For switch expressions and conditional expressions, we could potentially use function types in the best common type calculation of the overall expression type.

**Comment by jaredpar** [comment]
Moving to csharplang as this is a design question vs. compiler impl.

**Comment by CyrusNajmabadi** [comment]
Definitely seems odd.  I can't come up with a rational as to why array BCT works differently than switch or conditional BCT vis-a-vis function-types.

**Comment by HaloFour** [comment]
> I would also like to highlight another discrepancy related to functions.

IIRC, the team decided against supporting IIFE in those scenarios:
https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-07-26.md#direct-invocation-of-lambdas

**Comment by PrinceSatish** [comment]
By explicitly casting the lambdas to Func<string>, you provide the necessary type information for the compiler to determine the return type of the switch expression, and it will compile without errors.
